File: examples/python/hand/agg/handlers/post_blind.py
"""Handler for PostBlind command."""

import sys
from pathlib import Path
from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "angzarr"))
sys.path.insert(0, str(Path(__file__).parent.parent))

from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.errors import CommandRejectedError
from angzarr_client.proto.examples import hand_pb2 as hand
from angzarr_client import new_event_book

from .state import HandState


def handle_post_blind(
    command_book: types.CommandBook,
    command_any: Any,
    state: HandState,
    seq: int,
) -> types.EventBook:
    """Handle PostBlind command."""
    if not state.exists():
        raise CommandRejectedError("Hand not dealt")

    if state.status == "complete":
        raise CommandRejectedError("Hand is complete")

    cmd = hand.PostBlind()
    command_any.Unpack(cmd)

    if not cmd.player_root:
        raise CommandRejectedError("player_root is required")

    # Find the player
    player = None
    for p in state.players.values():
        if p.player_root == cmd.player_root:
            player = p
            break

    if not player:
        raise CommandRejectedError("Player not in hand")

    if player.has_folded:
        raise CommandRejectedError("Player has folded")

    if cmd.amount <= 0:
        raise CommandRejectedError("Blind amount must be positive")

    # Calculate actual amount (might be all-in)
    actual_amount = min(cmd.amount, player.stack)
    new_stack = player.stack - actual_amount
    new_pot_total = state.get_pot_total() + actual_amount

    event = hand.BlindPosted(
        player_root=cmd.player_root,
        blind_type=cmd.blind_type,
        amount=actual_amount,
        player_stack=new_stack,
        pot_total=new_pot_total,
        posted_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.googleapis.com/")

    return new_event_book(command_book, seq, event_any)

File: examples/rust/agg-hand/src/handlers/post_blind.rs
//! PostBlind command handler.

use angzarr_client::proto::examples::{BlindPosted, PostBlind};
use angzarr_client::proto::{event_page, CommandBook, EventBook, EventPage};
use angzarr_client::{pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::HandState;

pub fn handle_post_blind(
    command_book: &CommandBook,
    command_any: &Any,
    state: &HandState,
    seq: u32,
) -> CommandResult<EventBook> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Hand does not exist"));
    }
    if state.is_complete() {
        return Err(CommandRejectedError::new("Hand already complete"));
    }

    let cmd: PostBlind = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    let player = state
        .get_player(&cmd.player_root)
        .ok_or_else(|| CommandRejectedError::new("Player not in hand"))?;

    if cmd.amount <= 0 {
        return Err(CommandRejectedError::new("Amount must be positive"));
    }

    // Calculate actual amount (might be all-in)
    let actual_amount = cmd.amount.min(player.stack);
    let new_stack = player.stack - actual_amount;
    let new_pot = state.total_pot() + actual_amount;

    let event = BlindPosted {
        player_root: cmd.player_root,
        blind_type: cmd.blind_type,
        amount: actual_amount,
        player_stack: new_stack,
        pot_total: new_pot,
        posted_at: Some(angzarr_client::now()),
    };

    let event_any = pack_event(&event, "examples.BlindPosted");

    Ok(EventBook {
        cover: command_book.cover.clone(),
        pages: vec![EventPage {
            sequence: Some(event_page::Sequence::Num(seq)),
            event: Some(event_any),
            created_at: Some(angzarr_client::now()),
        }],
        snapshot: None,
        next_sequence: 0,
    })
}
